Fix Prefect deployment code (#1775)
kedro-org/kedro: docs/source/deployment/prefect.md

    flow.run_agent()

class KedroInitTask(Task):

>     """Task to initialize kedro session"""

Reviewer: ```suggestion
    """Task to initialize KedroSession"""
```

Author: Noticed that I haven't set up github to sign off on suggest changes I commit through the web interface. Will solve that problem some other time. For now I am implementing all your suggestions locally to allow for sign off.

    def __init__(
        self,
        pipeline_name: str,

>         package_name: str = None,

Reviewer: This shouldn't really be `None` though right? It wouldn't fail, but you'd get an odd name of "None_initialization" in the snippet below `f"{package_name}_initialization"`. I think you can just add it as a required argument.

Author: :+1:

        project_path: Union[Path, str] = None,
        env: str = None,
        extra_params: Dict[str, Any] = None,
        *args,
        **kwargs,
    ):
        self.project_path = Path(project_path or Path.cwd()).resolve()
        self.extra_params = extra_params
        self.pipeline_name = pipeline_name
        self.env = env
        super().__init__(name=f"{package_name}_initialization", *args, **kwargs)

    def run(self) -> Tuple[DataCatalog, KedroSession]:

>         """Initializes a kedro session and returns the data catalog and session"""

Reviewer: ```suggestion
        """Initializes a Kedro session and returns the DataCatalog and KedroSession"""
```

def instantiate_task(
    node: Node,
    tasks: Dict[str, Dict[str, Union[KedroTask, List[KedroTask]]]],
) -> Tuple[KedroTask, Dict[str, Dict[str, Union[KedroTask, List[KedroTask]]]]]:
    """
    Function pulls node task from <tasks> dictionary. If node task not available
    in <tasks> the function instantiates the tasks and adds it to <tasks>. In this
    way we avoid duplicate instantiations of the same node task.

    Args:
        node: kedro node for which a prefect task is being created.
        tasks: dictionary mapping node names to a dictionary containing
        node tasks and parent node tasks.

    Returns: prefect task for the passed node and task dictionary.

    """
    if tasks.get(node._unique_key) is not None:
        node_task = tasks[node._unique_key]["task"]
    else:
        node_task = KedroTask(node)
        tasks[node._unique_key] = {"task": node_task}

    # return tasks as it is mutated. We want to make this obvious to the user.
    return node_task, tasks  # type: ignore[return-value]

def generate_flow(
    init_task: KedroInitTask,
    tasks: Dict[str, Dict[str, Union[KedroTask, List[KedroTask]]]],
):
    """

>     Constructs a prefect flow given a task dictionary. Task dictionary

Reviewer: ```suggestion
    Constructs a Prefect flow given a task dictionary. Task dictionary
```

>     maps kedro node names to a dictionary containing a node task and its

Reviewer: ```suggestion
    maps Kedro node names to a dictionary containing a node task and its
```

    parents.

    Args:
        init_task: Prefect initialisation tasks. Used to instantiate a Kedro
        session within the scope of a Prefect flow.
        tasks: dictionary mapping Kedro node names to a dictionary
        containing a corresponding node task and its parents.

    Returns: None
    """

>     catalog, session = init_task

Reviewer 2: This is the weird bit that generated kedro_prefect_initialization[0] and kedro_prefect_intialization[1].

Not sure what is the correct thing, but it should be a `KedroInitTask` object instead and a session

    for task in tasks.values():
        node_task = task["task"]
        if len(task["parent_tasks"]) == 0:
            # When a task has no parent only the session init task should precede it.
            parent_tasks = [init_task]
        else:
            parent_tasks = task["parent_tasks"]
        # Set upstream tasks and bind required kwargs
        node_task.bind(upstream_tasks=parent_tasks, catalog=catalog, session=session)  # type: ignore[union-attr]

def instantiate_client(project_name: str):

>     """Initiates prefect client"""

Reviewer: ```suggestion
    """Initiates Prefect client"""
```
